File: ios/Classes/WatchtowerPlugin.m
#import "WatchtowerPlugin.h"
#import "WatchtowerSessionRecording.h"

@interface WatchtowerPlugin()
@property (nonatomic, strong) FlutterMethodChannel *methodChannel;
@property (nonatomic, strong) FlutterEventChannel *eventChannel;
@property (nonatomic, strong) WatchtowerSessionRecording *screenRecording;
@end

@implementation WatchtowerPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
    WatchtowerPlugin *instance = [[WatchtowerPlugin alloc] init];
    
    // Создаём MethodChannel
    instance.methodChannel = [FlutterMethodChannel
        methodChannelWithName:@"com.watchtower.plugin/screen_recording"
              binaryMessenger:[registrar messenger]];
    [registrar addMethodCallDelegate:instance channel:instance.methodChannel];
    
    // Создаём EventChannel
    instance.eventChannel = [FlutterEventChannel
        eventChannelWithName:@"com.watchtower.plugin/screen_recording_stream"
             binaryMessenger:[registrar messenger]];
    [instance.eventChannel setStreamHandler:instance.screenRecording];
    
    // Создаём экземпляр записи экрана
    instance.screenRecording = [[WatchtowerSessionRecording alloc] init];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([@"startRecorder" isEqualToString:call.method]) {
        NSNumber *interval = call.arguments[@"interval"];
        if (interval != nil) {
            [self.screenRecording startRecorder:[interval integerValue]];
            result(nil);
        } else {
            result([FlutterError errorWithCode:@"INVALID_ARGUMENT"
                                       message:@"Interval is required"
                                       details:nil]);
        }
    } else {
        result(FlutterMethodNotImplemented);
    }
}

@end

File: android/src/main/kotlin/com/kitegroup/watchtower/WatchtowerPlugin.kt
package com.kitegroup.watchtower

import android.annotation.SuppressLint
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class WatchtowerPlugin : FlutterPlugin, ActivityAware, MethodChannel.MethodCallHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var screenRecording: WatchtowerSessionRecording? = null
    private var eventSink: EventChannel.EventSink? = null

    companion object {
        private const val METHOD_CHANNEL = "com.watchtower.plugin/screen_recording"
        private const val EVENT_CHANNEL = "com.watchtower.plugin/screen_recording_stream"
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        ScreenRecorderInitializer()

        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL)
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })

        screenRecording = WatchtowerSessionRecording { frameData ->
            eventSink?.success(frameData)
        }
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startRecorder" -> {
                try {
                    val interval = call.argument<Int>("interval")
                    if (interval != null) {
                        screenRecording?.startRecorder(interval.toLong())
                        result.success(null)
                    } else {
                        result.error("INVALID_ARGUMENT", "Interval is required", null)
                    }
                } catch (e: Exception) {
                    result.error("ERROR", "Failed to start recorder: ${e.message}", null)
                }
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        screenRecording?.dispose()
        screenRecording = null
        eventSink = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    }

    override fun onDetachedFromActivityForConfigChanges() {
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    }

    override fun onDetachedFromActivity() {
    }
}
